File: clients/client-python/gravitino/client/generic_model_catalog.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import Dict, List

from gravitino.name_identifier import NameIdentifier
from gravitino.api.catalog import Catalog
from gravitino.api.model.model import Model
from gravitino.api.model.model_version import ModelVersion
from gravitino.client.base_schema_catalog import BaseSchemaCatalog
from gravitino.client.generic_model import GenericModel
from gravitino.client.generic_model_version import GenericModelVersion
from gravitino.dto.audit_dto import AuditDTO
from gravitino.dto.requests.model_register_request import ModelRegisterRequest
from gravitino.dto.requests.model_version_link_request import ModelVersionLinkRequest
from gravitino.dto.responses.base_response import BaseResponse
from gravitino.dto.responses.drop_response import DropResponse
from gravitino.dto.responses.entity_list_response import EntityListResponse
from gravitino.dto.responses.model_response import ModelResponse
from gravitino.dto.responses.model_version_list_response import ModelVersionListResponse
from gravitino.dto.responses.model_vesion_response import ModelVersionResponse
from gravitino.exceptions.handlers.model_error_handler import MODEL_ERROR_HANDLER
from gravitino.namespace import Namespace
from gravitino.rest.rest_utils import encode_string
from gravitino.utils import HTTPClient


class GenericModelCatalog(BaseSchemaCatalog):
    """
    The generic model catalog is a catalog that supports model and model version operations,
    for example, model register, model version link, model and model version list, etc.
    A model catalog is under the metalake.
    """

    def __init__(
        self,
        namespace: Namespace,
        name: str = None,
        catalog_type: Catalog.Type = Catalog.Type.UNSUPPORTED,
        provider: str = None,
        comment: str = None,
        properties: Dict[str, str] = None,
        audit: AuditDTO = None,
        rest_client: HTTPClient = None,
    ):
        super().__init__(
            namespace,
            name,
            catalog_type,
            provider,
            comment,
            properties,
            audit,
            rest_client,
        )

    def as_model_catalog(self):
        return self

    def list_models(self, namespace: Namespace) -> List[NameIdentifier]:
        """List the models in a schema namespace from the catalog.

        Args:
            namespace: The namespace of the schema.

        Raises:
            NoSuchSchemaException: If the schema does not exist.

        Returns:
            A list of NameIdentifier of models under the given namespace.
        """
        self._check_model_namespace(namespace)

        model_full_ns = self._model_full_namespace(namespace)
        resp = self.rest_client.get(
            self._format_model_request_path(model_full_ns),
            error_handler=MODEL_ERROR_HANDLER,
        )
        entity_list_resp = EntityListResponse.from_json(resp.body, infer_missing=True)
        entity_list_resp.validate()

        return [
            NameIdentifier.of(ident.namespace().level(2), ident.name())
            for ident in entity_list_resp.identifiers()
        ]

    def get_model(self, ident: NameIdentifier) -> Model:
        """Get a model by its identifier.

        Args:
            ident: The identifier of the model.

        Raises:
            NoSuchModelException: If the model does not exist.

        Returns:
            The model object.
        """
        self._check_model_ident(ident)

        model_full_ns = self._model_full_namespace(ident.namespace())
        resp = self.rest_client.get(
            f"{self._format_model_request_path(model_full_ns)}/{encode_string(ident.name())}",
            error_handler=MODEL_ERROR_HANDLER,
        )
        model_resp = ModelResponse.from_json(resp.body, infer_missing=True)
        model_resp.validate()

        return GenericModel(model_resp.model())

    def register_model(
        self, ident: NameIdentifier, comment: str, properties: Dict[str, str]
    ) -> Model:
        """Register a model in the catalog if the model is not existed, otherwise the
        ModelAlreadyExistsException will be thrown. The Model object will be created when the
        model is registered, users can call ModelCatalog#link_model_version to link the model
        version to the registered Model.

        Args:
            ident: The identifier of the model.
            comment: The comment of the model.
            properties: The properties of the model.

        Raises:
            ModelAlreadyExistsException: If the model already exists.
            NoSuchSchemaException: If the schema does not exist.

        Returns:
            The registered model object.
        """
        self._check_model_ident(ident)

        model_full_ns = self._model_full_namespace(ident.namespace())
        model_req = ModelRegisterRequest(
            name=encode_string(ident.name()), comment=comment, properties=properties
        )
        model_req.validate()

        resp = self.rest_client.post(
            self._format_model_request_path(model_full_ns),
            model_req,
            error_handler=MODEL_ERROR_HANDLER,
        )
        model_resp = ModelResponse.from_json(resp.body, infer_missing=True)
        model_resp.validate()

        return GenericModel(model_resp.model())

    def delete_model(self, model_ident: NameIdentifier) -> bool:
        """Delete the model from the catalog. If the model does not exist, return false.
        If the model is successfully deleted, return true. The deletion of the model will also
        delete all the model versions linked to this model.

        Args:
            model_ident: The identifier of the model.

        Returns:
            True if the model is deleted successfully, False is the model does not exist.
        """
        self._check_model_ident(model_ident)

        model_full_ns = self._model_full_namespace(model_ident.namespace())
        resp = self.rest_client.delete(
            f"{self._format_model_request_path(model_full_ns)}/{encode_string(model_ident.name())}",
            error_handler=MODEL_ERROR_HANDLER,
        )
        drop_resp = DropResponse.from_json(resp.body, infer_missing=True)
        drop_resp.validate()

        return drop_resp.dropped()

    def list_model_versions(self, model_ident: NameIdentifier) -> List[int]:
        """List all the versions of the register model by NameIdentifier in the catalog.

        Args:
            model_ident: The identifier of the model.

        Raises:
            NoSuchModelException: If the model does not exist.

        Returns:
            A list of model versions.
        """
        self._check_model_ident(model_ident)

        model_full_ident = self._model_full_identifier(model_ident)
        resp = self.rest_client.get(
            f"{self._format_model_version_request_path(model_full_ident)}/versions",
            error_handler=MODEL_ERROR_HANDLER,
        )
        model_version_list_resp = ModelVersionListResponse.from_json(
            resp.body, infer_missing=True
        )
        model_version_list_resp.validate()

        return model_version_list_resp.versions()

    def get_model_version(
        self, model_ident: NameIdentifier, version: int
    ) -> ModelVersion:
        """Get a model version by its identifier and version.

        Args:
            model_ident: The identifier of the model.
            version: The version of the model.

        Raises:
            NoSuchModelVersionException: If the model version does not exist.

        Returns:
            The model version object.
        """
        self._check_model_ident(model_ident)

        model_full_ident = self._model_full_identifier(model_ident)
        resp = self.rest_client.get(
            f"{self._format_model_version_request_path(model_full_ident)}/versions/{version}",
            error_handler=MODEL_ERROR_HANDLER,
        )
        model_version_resp = ModelVersionResponse.from_json(
            resp.body, infer_missing=True
        )
        model_version_resp.validate()

        return GenericModelVersion(model_version_resp.model_version())

    def get_model_version_by_alias(
        self, model_ident: NameIdentifier, alias: str
    ) -> ModelVersion:
        """
        Get a model version by its identifier and alias.

        Args:
            model_ident: The identifier of the model.
            alias: The alias of the model version.

        Raises:
            NoSuchModelVersionException: If the model version does not exist.

        Returns:
            The model version object.
        """
        self._check_model_ident(model_ident)

        model_full_ident = self._model_full_identifier(model_ident)
        resp = self.rest_client.get(
            f"{self._format_model_version_request_path(model_full_ident)}/aliases/"
            f"{encode_string(alias)}",
            error_handler=MODEL_ERROR_HANDLER,
        )
        model_version_resp = ModelVersionResponse.from_json(
            resp.body, infer_missing=True
        )
        model_version_resp.validate()

        return GenericModelVersion(model_version_resp.model_version())

    def link_model_version(
        self,
        model_ident: NameIdentifier,
        uri: str,
        aliases: List[str],
        comment: str,
        properties: Dict[str, str],
    ) -> None:
        """Link a new model version to the registered model object. The new model version will be
        added to the model object. If the model object does not exist, it will throw an
        exception. If the version alias already exists in the model, it will throw an exception.

        Args:
            model_ident: The identifier of the model.
            uri: The URI of the model version.
            aliases: The aliases of the model version. The aliases of the model version. The
            aliases should be unique in this model, otherwise the
            ModelVersionAliasesAlreadyExistException will be thrown. The aliases are optional and
            can be empty.
            comment: The comment of the model version.
            properties: The properties of the model version.

        Raises:
            NoSuchModelException: If the model does not exist.
            ModelVersionAliasesAlreadyExistException: If the aliases of the model version already exist.
        """
        self._check_model_ident(model_ident)

        model_full_ident = self._model_full_identifier(model_ident)

        request = ModelVersionLinkRequest(uri, comment, aliases, properties)
        request.validate()

        resp = self.rest_client.post(
            f"{self._format_model_version_request_path(model_full_ident)}/versions",
            request,
            error_handler=MODEL_ERROR_HANDLER,
        )
        base_resp = BaseResponse.from_json(resp.body, infer_missing=True)
        base_resp.validate()

    def delete_model_version(self, model_ident: NameIdentifier, version: int) -> bool:
        """Delete the model version from the catalog. If the model version does not exist, return false.
        If the model version is successfully deleted, return true.

        Args:
            model_ident: The identifier of the model.
            version: The version of the model.

        Returns:
            True if the model version is deleted successfully, False is the model version does not exist.
        """
        self._check_model_ident(model_ident)

        model_full_ident = self._model_full_identifier(model_ident)
        resp = self.rest_client.delete(
            f"{self._format_model_version_request_path(model_full_ident)}/versions/{version}",
            error_handler=MODEL_ERROR_HANDLER,
        )
        drop_resp = DropResponse.from_json(resp.body, infer_missing=True)
        drop_resp.validate()

        return drop_resp.dropped()

    def delete_model_version_by_alias(
        self, model_ident: NameIdentifier, alias: str
    ) -> bool:
        """Delete the model version by alias from the catalog. If the model version does not exist,
        return false. If the model version is successfully deleted, return true.

        Args:
            model_ident: The identifier of the model.
            alias: The alias of the model version.

        Returns:
            True if the model version is deleted successfully, False is the model version does not exist.
        """
        self._check_model_ident(model_ident)

        model_full_ident = self._model_full_identifier(model_ident)
        resp = self.rest_client.delete(
            f"{self._format_model_version_request_path(model_full_ident)}/aliases/"
            f"{encode_string(alias)}",
            error_handler=MODEL_ERROR_HANDLER,
        )
        drop_resp = DropResponse.from_json(resp.body, infer_missing=True)
        drop_resp.validate()

        return drop_resp.dropped()

    def register_model_version(
        self,
        ident: NameIdentifier,
        uri: str,
        aliases: List[str],
        comment: str,
        properties: Dict[str, str],
    ) -> Model:
        """Register a model in the catalog if the model is not existed, otherwise the
        ModelAlreadyExistsException will be thrown. The Model object will be created when the
        model is registered, in the meantime, the model version (version 0) will also be created and
        linked to the registered model. Register a model in the catalog and link a new model
        version to the registered model.

        Args:
            ident: The identifier of the model.
            uri: The URI of the model version.
            aliases: The aliases of the model version.
            comment: The comment of the model.
            properties: The properties of the model.

        Raises:
            ModelAlreadyExistsException: If the model already exists.
            ModelVersionAliasesAlreadyExistException: If the aliases of the model version already exist.

        Returns:
            The registered model object.
        """
        model = self.register_model(ident, comment, properties)
        self.link_model_version(ident, uri, aliases, comment, properties)
        return model

    def _check_model_namespace(self, namespace: Namespace):
        """Check the validity of the model namespace.

        Args:
            namespace: The namespace of the schema.

        Raises:
            IllegalNamespaceException: If the namespace is illegal.
        """
        Namespace.check(
            namespace is not None and namespace.length() == 1,
            f"Model namespace must be non-null and have 1 level, the input namespace is {namespace}",
        )

    def _check_model_ident(self, ident: NameIdentifier):
        """Check the validity of the model identifier.

        Args:
            ident: The identifier of the model.

        Raises:
            IllegalNameIdentifierException: If the identifier is illegal.
            IllegalNamespaceException: If the namespace is illegal.
        """
        NameIdentifier.check(
            ident is not None and ident.has_namespace(),
            f"Model identifier must be non-null and have a namespace, the input identifier is {ident}",
        )
        NameIdentifier.check(
            ident.name() is not None and len(ident.name()) > 0,
            f"Model name must be non-null and non-empty, the input name is {ident.name()}",
        )
        self._check_model_namespace(ident.namespace())

    def _format_model_request_path(self, model_ns: Namespace) -> str:
        """Format the model request path.

        Args:
            model_ns: The namespace of the model.

        Returns:
            The formatted model request path.
        """
        schema_ns = Namespace.of(model_ns.level(0), model_ns.level(1))
        return (
            f"{BaseSchemaCatalog.format_schema_request_path(schema_ns)}/"
            f"{encode_string(model_ns.level(2))}/models"
        )

    def _format_model_version_request_path(self, model_ident: NameIdentifier) -> str:
        """Format the model version request path.

        Args:
            model_ident: The identifier of the model.

        Returns:
            The formatted model version request path.
        """
        return (
            f"{self._format_model_request_path(model_ident.namespace())}"
            f"/{encode_string(model_ident.name())}"
        )

    def _model_full_namespace(self, model_namespace: Namespace) -> Namespace:
        """Get the full namespace of the model.

        Args:
            model_namespace: The namespace of the model.

        Returns:
            The full namespace of the model.
        """
        return Namespace.of(
            self._catalog_namespace.level(0), self.name(), model_namespace.level(0)
        )

    def _model_full_identifier(self, model_ident: NameIdentifier) -> NameIdentifier:
        """Get the full identifier of the model.

        Args:
            model_ident: The identifier of the model.

        Returns:
            The full identifier of the model.
        """
        return NameIdentifier.builder(
            self._model_full_namespace(model_ident.namespace()), model_ident.name()
        )

File: clients/client-java/src/main/java/org/apache/gravitino/client/GenericModelCatalog.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.apache.commons.lang3.StringUtils;
import org.apache.gravitino.Catalog;
import org.apache.gravitino.NameIdentifier;
import org.apache.gravitino.Namespace;
import org.apache.gravitino.dto.AuditDTO;
import org.apache.gravitino.dto.CatalogDTO;
import org.apache.gravitino.dto.requests.ModelRegisterRequest;
import org.apache.gravitino.dto.requests.ModelUpdateRequest;
import org.apache.gravitino.dto.requests.ModelUpdatesRequest;
import org.apache.gravitino.dto.requests.ModelVersionLinkRequest;
import org.apache.gravitino.dto.requests.ModelVersionUpdateRequest;
import org.apache.gravitino.dto.requests.ModelVersionUpdatesRequest;
import org.apache.gravitino.dto.responses.BaseResponse;
import org.apache.gravitino.dto.responses.DropResponse;
import org.apache.gravitino.dto.responses.EntityListResponse;
import org.apache.gravitino.dto.responses.ModelResponse;
import org.apache.gravitino.dto.responses.ModelVersionInfoListResponse;
import org.apache.gravitino.dto.responses.ModelVersionListResponse;
import org.apache.gravitino.dto.responses.ModelVersionResponse;
import org.apache.gravitino.dto.responses.ModelVersionUriResponse;
import org.apache.gravitino.exceptions.ModelAlreadyExistsException;
import org.apache.gravitino.exceptions.ModelVersionAliasesAlreadyExistException;
import org.apache.gravitino.exceptions.NoSuchModelException;
import org.apache.gravitino.exceptions.NoSuchModelVersionException;
import org.apache.gravitino.exceptions.NoSuchModelVersionURINameException;
import org.apache.gravitino.exceptions.NoSuchSchemaException;
import org.apache.gravitino.model.Model;
import org.apache.gravitino.model.ModelCatalog;
import org.apache.gravitino.model.ModelChange;
import org.apache.gravitino.model.ModelVersion;
import org.apache.gravitino.model.ModelVersionChange;
import org.apache.gravitino.rest.RESTUtils;

class GenericModelCatalog extends BaseSchemaCatalog implements ModelCatalog {

  GenericModelCatalog(
      Namespace namespace,
      String catalogName,
      Catalog.Type catalogType,
      String provider,
      String comment,
      Map<String, String> properties,
      AuditDTO auditDTO,
      RESTClient restClient) {
    super(namespace, catalogName, catalogType, provider, comment, properties, auditDTO, restClient);
  }

  @Override
  public ModelCatalog asModelCatalog() {
    return this;
  }

  @Override
  public NameIdentifier[] listModels(Namespace namespace) throws NoSuchSchemaException {
    checkModelNamespace(namespace);

    Namespace modelFullNs = modelFullNamespace(namespace);
    EntityListResponse resp =
        restClient.get(
            formatModelRequestPath(modelFullNs),
            EntityListResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return Arrays.stream(resp.identifiers())
        .map(id -> NameIdentifier.of(id.namespace().level(2), id.name()))
        .toArray(NameIdentifier[]::new);
  }

  @Override
  public Model getModel(NameIdentifier ident) throws NoSuchModelException {
    checkModelNameIdentifier(ident);

    Namespace modelFullNs = modelFullNamespace(ident.namespace());
    ModelResponse resp =
        restClient.get(
            formatModelRequestPath(modelFullNs) + "/" + RESTUtils.encodeString(ident.name()),
            ModelResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return new GenericModel(resp.getModel(), restClient, modelFullNs);
  }

  @Override
  public Model registerModel(NameIdentifier ident, String comment, Map<String, String> properties)
      throws NoSuchSchemaException, ModelAlreadyExistsException {
    checkModelNameIdentifier(ident);

    Namespace modelFullNs = modelFullNamespace(ident.namespace());
    ModelRegisterRequest req = new ModelRegisterRequest(ident.name(), comment, properties);
    req.validate();

    ModelResponse resp =
        restClient.post(
            formatModelRequestPath(modelFullNs),
            req,
            ModelResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
    return new GenericModel(resp.getModel(), restClient, modelFullNs);
  }

  @Override
  public boolean deleteModel(NameIdentifier ident) {
    checkModelNameIdentifier(ident);

    Namespace modelFullNs = modelFullNamespace(ident.namespace());
    DropResponse resp =
        restClient.delete(
            formatModelRequestPath(modelFullNs) + "/" + RESTUtils.encodeString(ident.name()),
            DropResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return resp.dropped();
  }

  @Override
  public int[] listModelVersions(NameIdentifier ident) throws NoSuchModelException {
    checkModelNameIdentifier(ident);

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    ModelVersionListResponse resp =
        restClient.get(
            formatModelVersionRequestPath(modelFullIdent) + "/versions",
            ModelVersionListResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return resp.getVersions();
  }

  @Override
  public ModelVersion[] listModelVersionInfos(NameIdentifier ident) throws NoSuchModelException {
    checkModelNameIdentifier(ident);

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    ModelVersionInfoListResponse resp =
        restClient.get(
            formatModelVersionRequestPath(modelFullIdent) + "/versions",
            ImmutableMap.of("details", "true"),
            ModelVersionInfoListResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return resp.getVersions();
  }

  @Override
  public ModelVersion getModelVersion(NameIdentifier ident, int version)
      throws NoSuchModelVersionException {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(version >= 0, "Model version must be non-negative");

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    ModelVersionResponse resp =
        restClient.get(
            formatModelVersionRequestPath(modelFullIdent) + "/versions/" + version,
            ModelVersionResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();
    return new GenericModelVersion(resp.getModelVersion());
  }

  @Override
  public ModelVersion getModelVersion(NameIdentifier ident, String alias)
      throws NoSuchModelVersionException {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(StringUtils.isNotBlank(alias), "Model alias must be non-empty");

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    ModelVersionResponse resp =
        restClient.get(
            formatModelVersionRequestPath(modelFullIdent)
                + "/aliases/"
                + RESTUtils.encodeString(alias),
            ModelVersionResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
    return new GenericModelVersion(resp.getModelVersion());
  }

  @Override
  public void linkModelVersion(
      NameIdentifier ident,
      Map<String, String> uris,
      String[] aliases,
      String comment,
      Map<String, String> properties)
      throws NoSuchModelException, ModelVersionAliasesAlreadyExistException {
    checkModelNameIdentifier(ident);

    ModelVersionLinkRequest req = new ModelVersionLinkRequest(uris, aliases, comment, properties);
    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    BaseResponse resp =
        restClient.post(
            formatModelVersionRequestPath(modelFullIdent) + "/versions",
            req,
            BaseResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
  }

  @Override
  public String getModelVersionUri(NameIdentifier ident, int version, String uriName)
      throws NoSuchModelVersionException, NoSuchModelVersionURINameException {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(version >= 0, "Model version must be non-negative");

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    Map<String, String> queryParam =
        uriName == null ? Collections.emptyMap() : ImmutableMap.of("uriName", uriName);
    ModelVersionUriResponse resp =
        restClient.get(
            formatModelVersionRequestPath(modelFullIdent) + "/versions/" + version + "/uri",
            queryParam,
            ModelVersionUriResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();
    return resp.getUri();
  }

  @Override
  public String getModelVersionUri(NameIdentifier ident, String alias, String uriName)
      throws NoSuchModelVersionException, NoSuchModelVersionURINameException {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(StringUtils.isNotBlank(alias), "Model alias must be non-empty");

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    Map<String, String> queryParam =
        uriName == null ? Collections.emptyMap() : ImmutableMap.of("uriName", uriName);
    ModelVersionUriResponse resp =
        restClient.get(
            formatModelVersionRequestPath(modelFullIdent)
                + "/aliases/"
                + RESTUtils.encodeString(alias)
                + "/uri",
            queryParam,
            ModelVersionUriResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
    return resp.getUri();
  }

  @Override
  public boolean deleteModelVersion(NameIdentifier ident, int version) {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(version >= 0, "Model version must be non-negative");

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    DropResponse resp =
        restClient.delete(
            formatModelVersionRequestPath(modelFullIdent) + "/versions/" + version,
            DropResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return resp.dropped();
  }

  @Override
  public boolean deleteModelVersion(NameIdentifier ident, String alias) {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(StringUtils.isNotBlank(alias), "Model alias must be non-empty");

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    DropResponse resp =
        restClient.delete(
            formatModelVersionRequestPath(modelFullIdent)
                + "/aliases/"
                + RESTUtils.encodeString(alias),
            DropResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());
    resp.validate();

    return resp.dropped();
  }

  @Override
  public Model alterModel(NameIdentifier ident, ModelChange... changes)
      throws NoSuchModelException, IllegalArgumentException {
    checkModelNameIdentifier(ident);

    // Convert ModelChange objects to DTO requests
    List<ModelUpdateRequest> updateRequests =
        Arrays.stream(changes)
            .map(DTOConverters::toModelUpdateRequest)
            .collect(Collectors.toList());

    ModelUpdatesRequest req = new ModelUpdatesRequest(updateRequests);
    req.validate();

    Namespace modelFullNs = modelFullNamespace(ident.namespace());
    ModelResponse resp =
        restClient.put(
            formatModelRequestPath(modelFullNs) + "/" + RESTUtils.encodeString(ident.name()),
            req,
            ModelResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
    return new GenericModel(resp.getModel(), restClient, modelFullNs);
  }

  @Override
  public ModelVersion alterModelVersion(
      NameIdentifier ident, int version, ModelVersionChange... changes)
      throws NoSuchModelException, NoSuchModelVersionException, IllegalArgumentException {
    checkModelNameIdentifier(ident);

    List<ModelVersionUpdateRequest> updateRequests =
        Arrays.stream(changes)
            .map(DTOConverters::toModelVersionUpdateRequest)
            .collect(Collectors.toList());

    ModelVersionUpdatesRequest req = new ModelVersionUpdatesRequest(updateRequests);
    req.validate();

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    ModelVersionResponse resp =
        restClient.put(
            formatModelVersionRequestPath(modelFullIdent) + "/versions/" + version,
            req,
            ModelVersionResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
    return new GenericModelVersion(resp.getModelVersion());
  }

  @Override
  public ModelVersion alterModelVersion(
      NameIdentifier ident, String alias, ModelVersionChange... changes)
      throws NoSuchModelException, IllegalArgumentException {
    checkModelNameIdentifier(ident);
    Preconditions.checkArgument(StringUtils.isNotBlank(alias), "Model alias must be non-empty");

    List<ModelVersionUpdateRequest> updateRequests =
        Arrays.stream(changes)
            .map(DTOConverters::toModelVersionUpdateRequest)
            .collect(Collectors.toList());

    ModelVersionUpdatesRequest req = new ModelVersionUpdatesRequest(updateRequests);
    req.validate();

    NameIdentifier modelFullIdent = modelFullNameIdentifier(ident);
    ModelVersionResponse resp =
        restClient.put(
            formatModelVersionRequestPath(modelFullIdent)
                + "/aliases/"
                + RESTUtils.encodeString(alias),
            req,
            ModelVersionResponse.class,
            Collections.emptyMap(),
            ErrorHandlers.modelErrorHandler());

    resp.validate();
    return new GenericModelVersion(resp.getModelVersion());
  }

  /** @return A new builder instance for {@link GenericModelCatalog}. */
  public static Builder builder() {
    return new Builder();
  }

  private void checkModelNamespace(Namespace namespace) {
    Namespace.check(
        namespace != null && namespace.length() == 1,
        "Model namespace must be non-null and only have 1 level, the input namespace is %s",
        namespace);
  }

  private void checkModelNameIdentifier(NameIdentifier ident) {
    NameIdentifier.check(ident != null, "Model name identifier must be non-null");
    NameIdentifier.check(
        StringUtils.isNotBlank(ident.name()), "Model name identifier must have a non-empty name");
    checkModelNamespace(ident.namespace());
  }

  private Namespace modelFullNamespace(Namespace modelNs) {
    return Namespace.of(catalogNamespace().level(0), name(), modelNs.level(0));
  }

  private NameIdentifier modelFullNameIdentifier(NameIdentifier modelIdent) {
    return NameIdentifier.of(modelFullNamespace(modelIdent.namespace()), modelIdent.name());
  }

  @VisibleForTesting
  static String formatModelRequestPath(Namespace modelNs) {
    Namespace schemaNs = Namespace.of(modelNs.level(0), modelNs.level(1));
    return new StringBuilder()
        .append(formatSchemaRequestPath(schemaNs))
        .append("/")
        .append(RESTUtils.encodeString(modelNs.level(2)))
        .append("/models")
        .toString();
  }

  @VisibleForTesting
  static String formatModelVersionRequestPath(NameIdentifier modelIdent) {
    return formatModelRequestPath(modelIdent.namespace())
        + "/"
        + RESTUtils.encodeString(modelIdent.name());
  }

  static class Builder extends CatalogDTO.Builder<Builder> {

    private RESTClient restClient;

    private Namespace namespace;

    private Builder() {}

    Builder withNamespace(Namespace namespace) {
      this.namespace = namespace;
      return this;
    }

    Builder withRestClient(RESTClient restClient) {
      this.restClient = restClient;
      return this;
    }

    @Override
    public GenericModelCatalog build() {
      Namespace.check(
          namespace != null && namespace.length() == 1,
          "Catalog namespace must be non-null and have 1 level, the input namespace is %s",
          namespace);
      Preconditions.checkArgument(StringUtils.isNotBlank(name), "name must not be blank");
      Preconditions.checkArgument(type != null, "type must not be null");
      Preconditions.checkArgument(StringUtils.isNotBlank(provider), "provider must not be blank");
      Preconditions.checkArgument(audit != null, "audit must not be null");
      Preconditions.checkArgument(restClient != null, "restClient must be set");

      return new GenericModelCatalog(
          namespace, name, type, provider, comment, properties, audit, restClient);
    }
  }
}
